cmd/kubeapps-apis/plugins/pkg/resources/namespaces.go

```diff
@@ -0,0 +1,144 @@
+	if err != nil {
+		if k8sErrors.IsForbidden(err) {
+			// The user doesn't have permissions to list namespaces, then use the current pod's service account
+			inClusterClient, err := inClusterClientGetter()
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "unable to get the in-cluster k8s client: '%v'", err)
+			}
+			namespaces, err = inClusterClient.CoreV1().Namespaces().List(backgroundCtx, metav1.ListOptions{})
+			if err != nil && k8sErrors.IsForbidden(err) {
+				// Not even the configured kubeapps-apis service account has permission
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+
+		// Filter namespaces in which the user has permissions to write (secrets) only
+		if namespaceList, err := filterAllowedNamespaces(typedClient, maxWorkers, namespaces.Items); err != nil {
+			return nil, err
+		} else {
+			return namespaceList, nil
+		}
+	} else {
+		// If the user can list namespaces, do not filter them
+		return namespaces.Items, nil
+	}
+}
+
+func nsCheckerWorker(client kubernetes.Interface, nsJobs <-chan checkNSJob, resultChan chan checkNSResult) {
+	for j := range nsJobs {
+		res, err := client.AuthorizationV1().SelfSubjectAccessReviews().Create(context.TODO(), &authorizationapi.SelfSubjectAccessReview{
+			Spec: authorizationapi.SelfSubjectAccessReviewSpec{
+				ResourceAttributes: &authorizationapi.ResourceAttributes{
+					Group:     "",
+					Resource:  "secrets",
+					Verb:      "get",
+					Namespace: j.ns.Name,
+				},
+			},
+		}, metav1.CreateOptions{})
+		var allowed bool
+		if err != nil {
+			allowed = false
+		} else {
+			allowed = res.Status.Allowed
+		}
+		resultChan <- checkNSResult{j, allowed, err}
+	}
+}
+
+// filterAllowedNamespaces check to which namespaces the user has access.
+// By access is considered the role of getting secrets in the namespace.
+func filterAllowedNamespaces(userClient kubernetes.Interface, maxWorkers int, namespaces []corev1.Namespace) ([]corev1.Namespace, error) {
+	var allowedNamespaces []corev1.Namespace
+
+	var wg sync.WaitGroup
+	workers := int(math.Min(float64(len(namespaces)), float64(maxWorkers)))
+	checkNSJobs := make(chan checkNSJob, workers)
+	nsCheckRes := make(chan checkNSResult, workers)
+
+	// Process maxReq ns at a time
+	for i := 0; i < workers; i++ {
+		wg.Add(1)
+		go func() {
+			nsCheckerWorker(userClient, checkNSJobs, nsCheckRes)
+			wg.Done()
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(nsCheckRes)
+	}()
+
+	go func() {
+		for _, ns := range namespaces {
+			checkNSJobs <- checkNSJob{ns}
+		}
+		close(checkNSJobs)
+	}()
+
+	// Start receiving results
+	for res := range nsCheckRes {
+		if res.Error == nil {
+			if res.allowed {
+				allowedNamespaces = append(allowedNamespaces, res.ns)
+			}
+		} else {
+			log.Errorf("failed to check namespace permissions. Got %v", res.Error)
+		}
+	}
+	return allowedNamespaces, nil
+}
+
+func FilterActiveNamespaces(namespaces []corev1.Namespace) []corev1.Namespace {
```

Comment:
I guess you're moving this logic as it will be re-used by the kapp plugin, great then.

Reply:
Indeed, it will be used by Kapp, Helm and Resources plugins.